warehouse/query-core/src/main/java/datawave/query/iterator/waitwindow/WaitWindowObserver.java

```diff
@@ -0,0 +1,417 @@
+package datawave.query.iterator.waitwindow;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.PartialKey;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.iterators.YieldCallback;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.io.Text;
+import org.apache.log4j.Logger;
+
+import datawave.query.attributes.Document;
+import datawave.query.attributes.DocumentKey;
+import datawave.query.attributes.WaitWindowExceededMetadata;
+import datawave.query.exceptions.WaitWindowOverrunException;
+import datawave.query.iterator.ResultCountingIterator;
+import datawave.query.iterator.profile.QuerySpan;
+
+// This class maintains common state and logic to determine if the QueryIterator's stack of boolean logic
```

Comment:
In general, it would be good to show examples of keys in the various states as appropriate for each of the functions below.  That might make it easier to follow.

Reply:
Comments added.